Fix GH-10370: File corruption in _php_stream_copy_to_stream_ex when using copy_file_range

nielsdos opened the pull request:
Fixes GH-10370

copy_file_range can return early without copying all the data. This is
legal behaviour and worked properly, unless the mmap fallback was used.
The mmap fallback would read too much data into the destination,
corrupting the destination file. Furthermore, if the mmap fallback would
fail and have to fallback to the regular file copying mechanism, a
similar issue would occur because both maxlen and haveread are modified.
Furthermore, there was a mmap-resource leak in one of the failure paths of
the mmap fallback code.
This patch fixes these issues. This also adds regression tests using the
new copy_file_range early-return simulation added in the previous
commit.

That being said, we should *also* look into repeating copy_file_range if the early-return happens because the operation was interrupted. But that should be a separate (feature) PR.

arnaud-lb reviewed (approved):
This looks good to me!

Could you add test variants that call [stream_copy_to_stream](https://www.php.net/stream_copy_to_stream) explicitly? So that these tests don't get noop'ed in case the implementation of PharData or file_put_contents change in the future.

nielsdos commented:
Thanks for you review.
I changed the test file to be filled with random bytes so the test is more strict and robust.
I also added the additional tests you requested.

arnaud-lb commented:
Thank you!
